### PR TITLE
Charspanents

### PR DESCRIPTION
**Pull request by Holado**
- Character spans now follow tokens through all changes
- Comprehensive character span tests for different token types
- Relies on changes in spanfix branches in [Tokenizer](https://github.com/mideind/Tokenizer/compare/spanfix?expand=1), [GreynirPackage](https://github.com/mideind/GreynirPackage/compare/spanfix?expand=1), and [GreynirCorrect](https://github.com/mideind/GreynirCorrect/compare/spanfix?expand=1).

### CODE REVIEW

File: correct.py [hunk 1]
```diff
@@ -153,7 +153,7 @@ def encode_sentence(sent: Sentence) -> Dict[str, Any]:
         nonlocal offset
         for ix, t in enumerate(sent.tokens):
             tokens[ix]["i"] = offset
-            offset += len(t.original or "")
+            offset += len(t.origin_spans or "")
```
Comment (vthorsteinsson) [inline correct.py:156]: Why is this better than `len(t.original or "")`?

File: nertokenizer.py [hunk 2]
```diff
@@ -143,16 +151,19 @@ def token_or_entity(token: Tok) -> Tok:
                 # Return an entity token with no definitions
                 # (this will eventually need to be looked up by full name when
                 # displaying or processing the article)
-                return token_ctor.Entity(token.txt)
+                new_ent = token_ctor.Entity("")
+                new_ent = new_ent.concatenate(token)
+                return new_ent
```
Comment (vthorsteinsson) [inline nertokenizer.py:156]: More compact: `return token_ctor.Entity("").concatenate(token)`

File: nertokenizer.py [hunk 3]
```diff
@@ -143,16 +151,19 @@ def token_or_entity(token: Tok) -> Tok:
...
             # Return the full name meanings
-            return token_ctor.Person(token.txt, tfull.person_names)
+            new_ent = token_ctor.Person("", tfull.person_names)
+            new_ent = new_ent.concatenate(token)
+            return new_ent
```
Comment (vthorsteinsson) [inline nertokenizer.py:160]: More compact: `return token_ctor.Person("", tfull.person_names).concatenate(token)`

File: nertokenizer.py [hunk 4]
```diff
@@ -301,3 +313,9 @@ def add_to_state(slist: List[str], entity: Entity) -> None:
     # print("\nLast names:\n{0}".format("\n".join("{0}: {1}".format(k, v) for k, v in lastnames.items())))
 
     assert not tq
+
+
+if __name__ == "__main__":
```
Comment (vthorsteinsson) [inline nertokenizer.py:318]: Delete debug (also delete `import pprint` at top of file)

File: tests/test_all.py [hunk 5]
```diff
@@ -43,6 +43,134 @@
 from nertokenizer import *  # noqa
 from settings import *  # noqa
 
+def checking(text, real):
+    resp = check_grammar(text)
+    i = list([ i["i"] for i in resp[0][0][0]['tokens'] ])
```
Comment (vthorsteinsson) [inline tests/test_all.py:48]: `list()` is unnecessary around a list comprehension

File: tests/test_all.py [hunk 6]
```diff
@@ -43,6 +43,134 @@
...
+    #    pprint(resp)
+    #else:
+    #    pprint(resp)
+
+
+def test_character_spans():
+
+    # Person token tests
+    text = "Á Clinton."
+    real = [0, 1, 8]
```
Comment (vthorsteinsson) [inline tests/test_all.py:63]: Is this correct? Shouldn't the array be [0, 1, 10]? (The 10 being the index of the period.)

File: tests/test_all.py [hunk 7]
```diff
@@ -43,6 +43,134 @@
...
+    checking(text, real)
+
+    text = "Charles Parkton."
+    real = [0, 15]
+    checking(text, real)
+    
+    text = "Hér er Nanna."
+    real = [0, 3, 5, 10]
```
Comment (vthorsteinsson) [inline tests/test_all.py:71]: [0, 3, 6, 11]?

File: tests/test_all.py (hunk 6 above)
Comment (Holado) [reply tests/test_all.py:63]: I've counted it so:

```
Á Clinton.
0 12345678

```
The period gets the index 8 here, according to the origin_spans count, that don't include the whitespace as such. 
If I change the code back to tracking t.original, I get [0, 1, 9], which seems to include the whitespace. 
I'll change it back and update the tests accordingly.

File: correct.py (hunk 1 above)
Comment (Holado) [reply correct.py:156]: This has been reverted and other code updated to fit that.

File: nertokenizer.py (hunk 2 above)
Comment (Holado) [reply nertokenizer.py:156]: Done

File: nertokenizer.py (hunk 3 above)
Comment (Holado) [reply nertokenizer.py:160]: Done

File: nertokenizer.py (hunk 4 above)
Comment (Holado) [reply nertokenizer.py:318]: Done

File: tests/test_all.py (hunk 5 above)
Comment (Holado) [reply tests/test_all.py:48]: Done

File: tests/test_all.py (hunk 6 above)
Comment (Holado) [reply tests/test_all.py:63]: Update done

File: tests/test_all.py (hunk 7 above)
Comment (Holado) [reply tests/test_all.py:71]: Update done

**Pull request merged**